pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/TimeSegmentPruner.java

```diff
@@ -0,0 +1,340 @@
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.DateTimeFieldSpec;
+import org.apache.pinot.spi.data.DateTimeFormatSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The {@code TimeSegmentPruner} prunes segments based on their time column start & end time metadata stored in ZK. The pruner
+ * supports queries with filter (or nested filter) of EQUALITY and RANGE predicates.
+ */
+public class TimeSegmentPruner implements SegmentPruner {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TimeSegmentPruner.class);
+  private static final long MAX_END_TIME = Long.MAX_VALUE;
+  private static final long MIN_START_TIME = 0;
+
+  private final String _tableNameWithType;
+  private final ZkHelixPropertyStore _propertyStore;
+  private final String _segmentZKMetadataPathPrefix;
+  private final String _timeColumn;
+  private final TimeUnit _timeUnit;
+
+  private volatile IntervalST _timeRangeMSToSegmentSearchTree;
+  private final Map _segmentToTimeRangeMSCache;
+
+  public TimeSegmentPruner(TableConfig tableConfig, ZkHelixPropertyStore<ZNRecord> propertyStore) {
+    _tableNameWithType = tableConfig.getTableName();
+    _propertyStore = propertyStore;
+    _segmentZKMetadataPathPrefix = ZKMetadataProvider.constructPropertyStorePathForResource(_tableNameWithType) + "/";
+    _timeColumn = tableConfig.getValidationConfig().getTimeColumnName();
+    Preconditions
+        .checkNotNull(_timeColumn, "Time column must be configured in table config for table: %s", _tableNameWithType);
+
+    Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType);
+    Preconditions.checkState(schema != null, "Failed to find schema for table: %s", _tableNameWithType);
+    DateTimeFieldSpec dateTimeSpec = schema.getSpecForTimeColumn(_timeColumn);
+    Preconditions.checkNotNull(dateTimeSpec, "Field spec must be specified in schema for time column: %s of table: %s",
+        _timeColumn, _tableNameWithType);
+    DateTimeFormatSpec formatSpec = new DateTimeFormatSpec(dateTimeSpec.getFormat());
+    _timeUnit = formatSpec.getColumnUnit();
+    Preconditions
+        .checkNotNull(_timeUnit, "Time unit must be configured in the field spec for time column: %s of table: %s",
+            _timeColumn, _tableNameWithType);
+    _segmentToTimeRangeMSCache = new HashMap<String, Interval>();
+  }
+
+  @Override
+  public void init(ExternalView externalView, IdealState idealState, Set<String> onlineSegments) {
+    onExternalViewChange(externalView, idealState, onlineSegments);
+  }
+
+  @Override
+  synchronized public void onExternalViewChange(ExternalView externalView, IdealState idealState, Set<String> onlineSegments) {
+    List<String> newSegments = new ArrayList<>();
+    for (String onlineSegment : onlineSegments) {
+      if (!_segmentToTimeRangeMSCache.containsKey(onlineSegment)) {
```

Comment:
Are we covering the following case?

1.`segment_0` gets uploaded with `start_time=10, end_time=12`
2. There was an issue with the offline pipeline, we refresh the `segment_0` with `start_time=10, end_time=11`.

In this case, `_segmentToTimeRangeMSCache` needs to be updated properly with the new interval. Otherwise, our pruner result will be wrong. I think that the partition pruner also does not currently handle this issue. We should at least add the `TODO` comment to handle this in the future.

One possible approach is to add TTL expiration to the cache to make sure the interval eventually gets updated to the correct value.

Reply:
Will the refresh call `refreshSegment()` on pruners? If yes, the cache will be updated.

Reply:
@snleee This is handled with the API `refreshSegment()`